src/write_log.c

```diff
@@ -56,21 +59,111 @@ static int wl_write_messages (const data_set_t *ds, const value_list_t *vl)
+    size_t bfill = 0;
+
+    if (0 != strcmp (ds->type, vl->type))
+    {
+        ERROR ("write_log plugin: DS type does not match value list type");
+        return -1;
+    }
+
+    format_json_initialize(buffer, &bfill, &bfree);
+    format_json_value_list(buffer, &bfill, &bfree, ds, vl,
+                           /* store rates = */ 0);
+    format_json_finalize(buffer, &bfill, &bfree);
+
+    INFO ("write_log values:\n%s", buffer);
+
+    return (0);
+} /* int wl_write_json */
 
 static int wl_write (const data_set_t *ds, const value_list_t *vl,
-        __attribute__ ((unused)) user_data_t *user_data)
+        user_data_t *user_data)
 {
-    int status;
+    int status = 0;
+    int mode = (int) (size_t) user_data->data;
 
-    status = wl_write_messages (ds, vl);
+    if (mode == WL_FORMAT_GRAPHITE)
+    {
+        status = wl_write_graphite (ds, vl);
+    }
+    else if (mode == WL_FORMAT_JSON)
+    {
+        status = wl_write_json (ds, vl);
+    }
 
     return (status);
 }
 
+static int wl_config (oconfig_item_t *ci) /* {{{ */
+{
+    int mode = 0;
+    for (int i = 0; i < ci->children_num; i++)
+    {
+        oconfig_item_t *child = ci->children + i;
+
+        if (strcasecmp ("Format", child->key) == 0)
+        {
+            char *mode_str = NULL;
+            if ((child->values_num != 1)
+                || (child->values[0].type != OCONFIG_TYPE_STRING))
+            {
+                ERROR ("write_log plugin: Option `%s' requires "
+                    "exactly one string argument.", child->key);
+                return (-EINVAL);
+            }
+            if (mode != 0)
+            {
+                WARNING ("write_log plugin: Redefining option `%s'.",
+                    child->key);
+            }
+            mode_str = child->values[0].value.string;
+            if (strcasecmp ("Graphite", mode_str) == 0)
+                mode = WL_FORMAT_GRAPHITE;
+            else if (strcasecmp ("JSON", mode_str) == 0)
+                mode = WL_FORMAT_JSON;
+            else
+            {
+                ERROR ("write_log plugin: Unknown mode `%s' for option `%s'.",
+                    mode_str, child->key);
+                return (-EINVAL);
+            }
+        }
+        else
+        {
+            ERROR ("write_log plugin: Invalid configuration option: `%s'.",
+                child->key);
+        }
+    }
+    if (mode == 0)
+        mode = WL_FORMAT_GRAPHITE;
+
+    user_data_t ud = {
+        .data = (void *) (size_t) mode,
+        .free_func = NULL
+    };
+
+    plugin_register_write ("write_log", wl_write, &ud);
+
+    return (0);
+} /* }}} int wl_config */
+
 void module_register (void)
 {
-    plugin_register_write ("write_log", wl_write, NULL);
+    plugin_register_complex_config ("write_log", wl_config);
+
+    user_data_t ud = {
+        .data = (void *) (size_t) WL_FORMAT_GRAPHITE,
+        .free_func = NULL
+    };
+
+    plugin_register_write ("write_log", wl_write, &ud);
```

Comment:
Thanks for the review! I'd like to understand a bit more about the backwards incompatibility here. The reason for the two `plugin_register_write()` calls is exactly to allow the plugin to run with or without a config block. The code here actually works, but does log an error when the callback is overridden (so previous behavior is completely unchanged). Is there an implied invariant that `plugin_register_write()` will always be called exactly once? Instead of a global variable, would it make sense to add some sort of a callback update function that effectively modifies the `user_data_t` of an existing callback?